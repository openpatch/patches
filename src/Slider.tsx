import { ChangeEvent, KeyboardEvent, ReactNode, useRef } from "react";
import { css } from "@emotion/react";
import { Box } from "./Box";

export type SliderProps = {
  id?: string;
  value?: number;
  onChange?: (n: number | null) => void;
  onBlur?: () => void;
  disabled?: boolean;
  required?: boolean;
  step?: number;
  labelLeft?: ReactNode;
  labelRight?: ReactNode;
  showValue?: boolean;
  min?: number;
  max?: number;
};

export const Slider = ({
  id,
  labelLeft = null,
  labelRight = null,
  value = 0,
  step = 1,
  min = 0,
  max = 100,
  disabled = false,
  required = false,
  showValue = true,
  onBlur = () => {},
  onChange = () => {},
}: SliderProps) => {
  const bubbleRef = useRef<HTMLDivElement>(null);

  function showBubble() {
    if (bubbleRef.current) {
      bubbleRef.current.style.visibility = "visible";
      bubbleRef.current.style.opacity = "1";
    }
  }

  function hideBubble() {
    if (bubbleRef.current) {
      bubbleRef.current.style.visibility = "hidden";
      bubbleRef.current.style.opacity = "0";
    }
  }

  function handleFocus() {
    showBubble();
  }

  function handleBlur() {
    onBlur();
    hideBubble();
  }

  function handleKeyDown(e: KeyboardEvent<HTMLInputElement>) {
    let flag = true;
    switch (e.key) {
      case "ArrowUp":
        increase();
        break;
      case "ArrowDown":
        decrease();
        break;
      case "Home":
        toMin();
        break;
      case "End":
        toMax();
        break;
      default:
        flag = false;
    }

    if (flag) {
      e.stopPropagation();
      e.preventDefault();
    }
  }

  function handleInnerChange(v: number | null) {
    if (!disabled) {
      onChange(v);
    }
  }

  function handleChange(e: ChangeEvent<HTMLInputElement>) {
    const n = Number(e.target.value);
    if (Number.isNaN(n)) {
      handleInnerChange(null);
    } else {
      handleInnerChange(n);
    }
  }

  function toMax() {
    if (max) {
      handleInnerChange(max);
    }
  }

  function toMin() {
    if (min) {
      handleInnerChange(min);
    }
  }

  function increase() {
    const newValue = value + step;
    if (!max || newValue <= max) {
      handleInnerChange(newValue);
    }
  }

  function decrease() {
    const newValue = value - step;
    if (!min || newValue >= min) {
      handleInnerChange(newValue);
    }
  }

  const thumbPosition = Number(((value - min) * 100) / (max - min));

  return (
    <Box display="flex" alignItems="center" justifyContent="center">
      <Box mr="xsmall">{labelLeft ? labelLeft : min}</Box>
      <Box position="relative" display="flex" flex={1} alignItems="center">
        <input
          id={id}
          onFocus={handleFocus}
          onBlur={handleBlur}
          onPointerOver={handleFocus}
          onPointerLeave={handleBlur}
          onPointerDown={handleFocus}
          onPointerUp={handleBlur}
          required={required}
          disabled={disabled}
          css={(theme) => [
            css`
              width: 100%;
              appearance: none;
              cursor: pointer;
              height: 8px;
              outline: none;
              border-radius: ${theme.radii.small};
              background-color: ${theme.colors.neutral["200"]};

              ::-moz-range-thumb,
              ::-webkit-slider-thumb {
                appearance: none;
                border: none;
                height: 20px;
                width: 20px;
                border-radius: 50%;
                background-color: ${theme.colors.primary["800"]};
              }
            `,
            disabled &&
              css`
                cursor: not-allowed;
                ::-moz-range-thumb,
                ::-webkit-slider-thumb {
                  background-color: ${theme.colors.neutral["200"]};
                }
              `,
          ]}
          type="range"
          min={min}
          max={max}
          step={step}
          value={value}
          onChange={handleChange}
          onKeyDown={handleKeyDown}
        />
        {showValue && (
          <output
            ref={bubbleRef}
            style={{
              left: `calc(${thumbPosition}% + (${
                10 - thumbPosition * 0.18
              }px))`,
            }}
            css={() => css`
              position: absolute;
              top: -36px;
              margin-left: -17px;
              z-index: 1;
              font-size: 0.75rem;
              visibility: hidden;
              opacity: 0;
              transform: scale(1) translateY(-10px);
              pointer-events: none;
              transition: visibility 0.33s linear, opacity 0.33s linear;
            `}
          >
            <span
              css={(theme) => [
                css`
                  width: 32px;
                  height: 32px;
                  display: flex;
                  transform: rotate(-45deg);
                  align-items: center;
                  border-radius: 50% 50% 50% 0;
                  justify-content: center;
                  background-color: ${theme.colors.primary["800"]};
                `,
                disabled &&
                  css`
                    background-color: ${theme.colors.neutral["200"]};
                  `,
              ]}
            >
              <span
                css={(theme) => css`
                  color: ${theme.colors.primary["50"]};
                  transform: rotate(45deg);
                `}
              >
                {value}
              </span>
            </span>
          </output>
        )}
      </Box>
      <Box ml="xsmall">{labelRight ? labelRight : max}</Box>
    </Box>
  );
};
