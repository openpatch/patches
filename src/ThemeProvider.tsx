import React, { ReactElement } from "react";
import { ThemeProvider as EThemeProvider } from "emotion-theming";
import { Theme, base } from "./themes";
import { GlobalCssTheme } from "./themes/utils";

interface ThemeProviderProps {
  theme?: Theme;
  children: ReactElement;
}

const ThemeProvider = ({ theme = base, children }: ThemeProviderProps) => {
  return (
    <EThemeProvider theme={theme}>
      <GlobalCssTheme theme={theme} />
      {children}
    </EThemeProvider>
  );
};

export default ThemeProvider;
