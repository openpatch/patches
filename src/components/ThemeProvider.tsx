import React, { ReactElement } from "react";
import {
  ThemeProvider as SCThemeProvider,
  DefaultTheme,
} from "styled-components";
import defaultTheme from "../shared/theme";

interface ThemeProviderProps {
  theme?: DefaultTheme;
  children: ReactElement;
}

const ThemeProvider = ({
  theme = defaultTheme,
  children,
}: ThemeProviderProps) => {
  return <SCThemeProvider theme={theme}>{children}</SCThemeProvider>;
};

export default ThemeProvider;
