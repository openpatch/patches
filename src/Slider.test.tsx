
import { render } from "@testing-library/react";
import { Slider, SliderProps } from "./Slider";
import { ThemeProvider } from "./ThemeProvider";

const renderWithTheme = (props?: SliderProps) =>
render(
    <ThemeProvider>
        <Slider {...props} />
    </ThemeProvider>
);

test("should render", () => {
   renderWithTheme();
});

test("should match snapshot", () => {
    const { container } = renderWithTheme();
    expect(container).toMatchSnapshot();
});