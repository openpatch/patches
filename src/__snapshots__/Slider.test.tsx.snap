// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
.emotion-0 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  scroll-behavior: smooth;
  line-height: 1.5;
  -webkit-text-size-adjust: 100%;
  color: #1F2933;
  letter-spacing: 0;
  font-size: 16px;
  font-family: Rubik,system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",sans-serif;
}

.emotion-0 h1 {
  font-size: 32px;
  margin: 0.67em 0;
}

.emotion-0 h1,
.emotion-0 h2,
.emotion-0 h3,
.emotion-0 h4,
.emotion-0 h5,
.emotion-0 h6 {
  font-family: Rubik,system-ui;
}

.emotion-0 hr {
  box-sizing: content-box;
  height: 0;
  overflow: visible;
}

.emotion-0 pre {
  font-family: "Ubuntu Mono",Menlo,monospace;
  font-size: 1em;
}

.emotion-0 a {
  background-color: transparent;
}

.emotion-0 abbr[title] {
  border-bottom: none;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  -webkit-text-decoration: underline dotted;
  text-decoration: underline dotted;
}

.emotion-0 b,
.emotion-0 strong {
  font-weight: bolder;
}

.emotion-0 code,
.emotion-0 kbd,
.emotion-0 samp {
  font-family: "Ubuntu Mono",Menlo,monospace;
  font-size: 1em;
}

.emotion-0 small {
  font-size: 80%;
}

.emotion-0 sub,
.emotion-0 sup {
  font-size: 75%;
  line-height: 0;
  position: relative;
  vertical-align: baseline;
}

.emotion-0 sub {
  bottom: -0.25em;
}

.emotion-0 sup {
  top: -0.5em;
}

.emotion-0 img {
  max-width: 100%;
  border-style: none;
}

.emotion-0 button,
.emotion-0 input,
.emotion-0 optgroup,
.emotion-0 select,
.emotion-0 textarea {
  font-family: inherit;
  font-size: 100%;
  line-height: 1.15;
  margin: 0;
}

.emotion-0 button,
.emotion-0 input {
  overflow: visible;
}

.emotion-0 button,
.emotion-0 select {
  text-transform: none;
}

.emotion-0 button,
.emotion-0 [type="button"],
.emotion-0 [type="reset"],
.emotion-0 [type="submit"] {
  -webkit-appearance: button;
}

.emotion-0 button::-moz-focus-inner,
.emotion-0 [type="button"]::-moz-focus-inner,
.emotion-0 [type="reset"]::-moz-focus-inner,
.emotion-0 [type="submit"]::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.emotion-0 button:-moz-focusring,
.emotion-0 [type="button"]:-moz-focusring,
.emotion-0 [type="reset"]:-moz-focusring,
.emotion-0 [type="submit"]:-moz-focusring {
  outline: 1px dotted ButtonText;
}

.emotion-0 fieldset {
  padding: 0.35em 0.75em 0.625em;
}

.emotion-0 legend {
  box-sizing: border-box;
  color: inherit;
  display: table;
  max-width: 100%;
  padding: 0;
  white-space: normal;
}

.emotion-0 progress {
  vertical-align: baseline;
}

.emotion-0 textarea {
  overflow: auto;
}

.emotion-0 [type="checkbox"],
.emotion-0 [type="radio"] {
  box-sizing: border-box;
  padding: 0;
}

.emotion-0 [type="number"]::-webkit-inner-spin-button,
.emotion-0 [type="number"]::-webkit-outer-spin-button {
  height: auto;
}

.emotion-0 [type="search"] {
  -webkit-appearance: textfield;
  outline-offset: -2px;
}

.emotion-0 [type="search"]::-webkit-search-decoration {
  -webkit-appearance: none;
}

.emotion-0::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}

.emotion-0 details {
  display: block;
}

.emotion-0 summary {
  display: -webkit-box;
  display: -webkit-list-item;
  display: -ms-list-itembox;
  display: list-item;
}

.emotion-0 template {
  display: none;
}

.emotion-0 [hidden] {
  display: none;
}

.emotion-1 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-2 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  margin-right: 8px;
}

.emotion-3 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-4 {
  width: 100%;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  cursor: pointer;
  height: 8px;
  outline: none;
  border-radius: 4px;
  background-color: #CBD2D9;
}

.emotion-4::-moz-range-thumb,
.emotion-4::-webkit-slider-thumb {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  border: none;
  height: 20px;
  width: 20px;
  border-radius: 50%;
  background-color: #0C6B58;
}

.emotion-5 {
  position: absolute;
  top: -36px;
  margin-left: -17px;
  z-index: 1;
  font-size: 0.75rem;
  visibility: hidden;
  opacity: 0;
  -webkit-transform: scale(1) translateY(-10px);
  -moz-transform: scale(1) translateY(-10px);
  -ms-transform: scale(1) translateY(-10px);
  transform: scale(1) translateY(-10px);
  pointer-events: none;
  -webkit-transition: visibility 0.33s linear,opacity 0.33s linear;
  transition: visibility 0.33s linear,opacity 0.33s linear;
}

.emotion-6 {
  width: 32px;
  height: 32px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-transform: rotate(-45deg);
  -moz-transform: rotate(-45deg);
  -ms-transform: rotate(-45deg);
  transform: rotate(-45deg);
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 50% 50% 50% 0;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  background-color: #0C6B58;
}

.emotion-7 {
  color: #EFFCF6;
  -webkit-transform: rotate(45deg);
  -moz-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
}

.emotion-8 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  margin-left: 8px;
}

<div>
  <div
    class="emotion-0"
    id="patches-root"
  >
    <div
      class="emotion-1"
      display="flex"
    >
      <div
        class="emotion-2"
      >
        0
      </div>
      <div
        class="emotion-3"
        display="flex"
      >
        <input
          class="emotion-4"
          max="100"
          min="0"
          step="1"
          type="range"
          value="0"
        />
        <output
          class="emotion-5"
        >
          <span
            class="emotion-6"
          >
            <span
              class="emotion-7"
            >
              0
            </span>
          </span>
        </output>
      </div>
      <div
        class="emotion-8"
      >
        100
      </div>
    </div>
  </div>
</div>
`;
